src/pybind/mgr/cephadm/serve.py

```diff
@@ -633,10 +633,10 @@ def _apply_service(self, spec: ServiceSpec) -> bool:
         self._apply_service_config(spec)
 
         if service_type == 'osd':
-            self.mgr.osd_service.create_from_spec(cast(DriveGroupSpec, spec))
-            # TODO: return True would result in a busy loop
-            # can't know if daemon count changed; create_from_spec doesn't
-            # return a solid indication
+            osd_svc_spec = cast(DriveGroupSpec, spec)
+            osd_distribution = self.mgr.osd_service.create_from_spec(osd_svc_spec)
+            osd_svc_spec.update_daemon_distribution(osd_distribution)
```

Comment:
Will this work with upgrades? As in, if I already had a few osds in a service and then upgraded to a cluster that uses this daemon distribution for osd service size would the old osds get into the daemon_distribution object for the service? It looks like this updates it with newly created osds but will we also add pre-existing ones?

Reply:
My view is that this is going to be called when deploying the new osd daemons with the new image so is going to work in the same way that is working for new services or services with changes. 
Can you confirm that upgrade a daemon is going to call "_apply_service" ?

Reply:
Upgrading does not call `_apply_service`. We just call `_daemon_action` with the action being "redeploy" and the image being the new image we're upgrading to. However, `_apply_service` is called for every service spec every time we go through the normal serve loop anyway so it might just work fine.

Reply:
I tried testing this and saw issues. Before upgrading, the size on my "defaultDG" and "all-available-devices" osds were listed with sizes of 9 and 8 (which is wrong, the actual sizes were 6 and 5).

```
[ceph: root@vm-00 /]# ceph orch ls
NAME                       PORTS  RUNNING  REFRESHED  AGE  PLACEMENT          
alertmanager                          1/1  23s ago    19m  count:1            
crash                                 3/3  101s ago   19m  *                  
grafana                               1/1  23s ago    19m  count:1            
mgr                                   2/2  101s ago   19m  count:2            
mon                                   3/5  101s ago   19m  count:5            
node-exporter                         3/3  101s ago   19m  *                  
osd.all-available-devices             5/8  101s ago   2m   *                  
osd.defaultDG                         6/9  101s ago   8m   vm-00;vm-01;vm-02  
osd.unmanaged                         4/4  101s ago   -    <unmanaged>        
prometheus                            1/1  23s ago    19m  count:1            
```

After upgrading to a cluster with these changes the size in both services was listed as 0

```
[ceph: root@vm-00 /]# ceph orch ls                                            
NAME                       PORTS  RUNNING  REFRESHED  AGE  PLACEMENT          
alertmanager                          1/1  7s ago     31m  count:1            
crash                                 3/3  8s ago     31m  *                  
grafana                               1/1  7s ago     31m  count:1            
mgr                                   2/2  8s ago     31m  count:2            
mon                                   3/5  8s ago     31m  count:5            
node-exporter                         3/3  8s ago     31m  *                  
osd.all-available-devices             5/0  8s ago     14m  *                  
osd.defaultDG                         6/0  8s ago     20m  vm-00;vm-01;vm-02  
osd.unmanaged                         4/4  8s ago     -    <unmanaged>        
prometheus                            1/1  7s ago     31m  count:1       
```

However, if I re-applied the services the size was fixed

```
[ceph: root@vm-00 /]# ceph orch ls                                            
NAME                       PORTS  RUNNING  REFRESHED  AGE  PLACEMENT          
alertmanager                          1/1  7s ago     31m  count:1            
crash                                 3/3  8s ago     31m  *                  
grafana                               1/1  7s ago     31m  count:1            
mgr                                   2/2  8s ago     31m  count:2            
mon                                   3/5  8s ago     31m  count:5            
node-exporter                         3/3  8s ago     31m  *                  
osd.all-available-devices             5/0  8s ago     14m  *                  
osd.defaultDG                         6/0  8s ago     20m  vm-00;vm-01;vm-02  
osd.unmanaged                         4/4  8s ago     -    <unmanaged>        
prometheus                            1/1  7s ago     31m  count:1            
[ceph: root@vm-00 /]# ceph orch apply osd --all-available-devices
Scheduled osd.all-available-devices update...
[ceph: root@vm-00 /]# ceph orch ls
NAME                       PORTS  RUNNING  REFRESHED  AGE  PLACEMENT          
alertmanager                          1/1  4m ago     36m  count:1            
crash                                 3/3  4m ago     36m  *                  
grafana                               1/1  4m ago     36m  count:1            
mgr                                   2/2  4m ago     36m  count:2            
mon                                   3/5  4m ago     36m  count:5            
node-exporter                         3/3  4m ago     36m  *                  
osd.all-available-devices             5/5  4m ago     55s  *                  
osd.defaultDG                         6/0  4m ago     24m  vm-00;vm-01;vm-02  
osd.unmanaged                         4/4  4m ago     -    <unmanaged>        
prometheus                            1/1  4m ago     36m  count:1            
[ceph: root@vm-00 /]# ceph orch apply -i osd_spec1.yaml 
Scheduled osd.defaultDG update...
[ceph: root@vm-00 /]# ceph orch ls
NAME                       PORTS  RUNNING  REFRESHED  AGE  PLACEMENT          
alertmanager                          1/1  5m ago     37m  count:1            
crash                                 3/3  5m ago     37m  *                  
grafana                               1/1  5m ago     37m  count:1            
mgr                                   2/2  5m ago     37m  count:2            
mon                                   3/5  5m ago     37m  count:5            
node-exporter                         3/3  5m ago     37m  *                  
osd.all-available-devices             5/5  5m ago     2m   *                  
osd.defaultDG                         6/6  5m ago     26s  vm-00;vm-01;vm-02  
osd.unmanaged                         4/4  5m ago     -    <unmanaged>        
prometheus                            1/1  5m ago     37m  count:1            
```
so I feel like this is very close to working properly. Maybe just some minor change to handle upgrading old osd services.

Reply:
Thank you for testing this change Adam... lest see how can we force a refresh after updating.. any idea?

Reply:
After a successful upgrade, I have forced the update of the osd services information in the manager cache, cleaning the cache dict "osdspec_last_applied".  
In this way we have an update of the total number of OSd daemons in each host for each osd service.